appservice/src/main/java/nu/yona/server/device/rest/DeviceController.java

```diff
@@ -0,0 +1,161 @@
+import org.springframework.hateoas.mvc.ControllerLinkBuilder;
+import org.springframework.hateoas.mvc.ResourceAssemblerSupport;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import nu.yona.server.crypto.seckey.CryptoSession;
+import nu.yona.server.device.rest.DeviceController.DeviceResource;
+import nu.yona.server.device.service.DeviceBaseDto;
+import nu.yona.server.device.service.DeviceService;
+import nu.yona.server.rest.JsonRootRelProvider;
+import nu.yona.server.subscriptions.rest.BuddyController;
+import nu.yona.server.subscriptions.service.UserService;
+
+@Controller
+@ExposesResourceFor(DeviceResource.class)
+@RequestMapping(value = "/users/{userId}/devices", produces = { MediaType.APPLICATION_JSON_VALUE })
+public class DeviceController
+{
+	@Autowired
+	private DeviceService deviceService;
+
+	@Autowired
+	private UserService userService;
+
+	@Autowired
+	private CurieProvider curieProvider;
+
+	@RequestMapping(value = "/", method = RequestMethod.GET)
+	@ResponseBody
+	public HttpEntity<Resources<DeviceResource>> getAllDevices(@RequestHeader(value = PASSWORD_HEADER) Optional<String> password,
+			@PathVariable UUID userId)
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(password, () -> userService.canAccessPrivateData(userId)))
+		{
+			return new ResponseEntity<>(
+					createAllDevicesCollectionResource(curieProvider, userId, deviceService.getDevicesOfUser(userId)),
+					HttpStatus.OK);
+		}
+	}
+
+	@RequestMapping(value = "/{buddyId}", method = RequestMethod.GET)
+	@ResponseBody
+	public HttpEntity<DeviceResource> getDevice(@RequestHeader(value = PASSWORD_HEADER) Optional<String> password,
+			@PathVariable UUID userId, @PathVariable UUID deviceId)
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(password, () -> userService.canAccessPrivateData(userId)))
+		{
+
+			return createOkResponse(userId, deviceService.getDevice(deviceId));
+		}
+	}
+
+	private HttpEntity<DeviceResource> createOkResponse(UUID userId, DeviceBaseDto device)
+	{
+		return createResponse(userId, device, HttpStatus.OK);
+	}
+
+	private HttpEntity<DeviceResource> createResponse(UUID userId, DeviceBaseDto device, HttpStatus status)
+	{
+		return new ResponseEntity<>(new DeviceResourceAssembler(curieProvider, userId).toResource(device), status);
+	}
+
+	public static Resources<DeviceResource> createAllDevicesCollectionResource(CurieProvider curieProvider, UUID userId,
+			Set<DeviceBaseDto> allBuddiesOfUser)
```

Comment:
`allBuddiesOfUser` should be `allDevicesOfUser`

Reply:
This one is already removed. It seems you're reviewing an older commit.